659 Exchange Staking Integration (#679)
casper-network/docs: source/docs/casper/dapp-dev-guide/list-cspr.md

## Understanding Basic Transactions

We have a token and transaction model that features different levels of support that range from convenient to robust. Usually, when you are transferring Casper tokens between two parties, the native two-party transfer will suffice.

> Casper supports native two-party transfers as well as bulk transfers using custom Wasm. The native transfer is ideal when you need to perform a one-to-one transfer between two accounts. Whereas, the batched Wasm transfer can be used when you are making bulk transfers. A batched Wasm transfer allows you to do multiple transfers in a single deploy, which makes it a more cost effective method when sending tokens from one account to several others.

Reviewer: ```suggestion
Casper supports native two-party transfers as well as bulk transfers using custom Wasm. The native transfer is ideal when you need to perform a one-to-one transfer between two accounts. Whereas the batched Wasm transfer can be used when making bulk transfers. A batched Wasm transfer allows you to do multiple transfers in a single deploy, making it more cost-effective when sending tokens from one account to several others.
```

### Native transfer

> Native transfers can be accomplished by sending a native transfer deploy, without any Wasm. Included below is an example of this type of deploy. The included `payment` field describes how we are paying for the deploy, in this case a native transfer, while the `session` field describes the actual transfer.

Reviewer: Is there time to move some of the passive voice to active voice? I notice several spots where the article uses passive voice (in the original text as well).

- [Delegate](https://github.com/casper-network/casper-node/tree/dev/smart_contracts/contracts/client/delegate)
- [Undelegate](https://github.com/casper-network/casper-node/tree/dev/smart_contracts/contracts/client/undelegate)
- [Redelegate](https://github.com/casper-network/casper-node/tree/dev/smart_contracts/contracts/client/redelegate)

#### 1. Creating a deploy object

To create a deploy using the JavaScript SDK, we will need `deployParams`, `session` and a `payment`.

Deploy params is a `DeployUtil.DeployParams` object created from the delegator's `publicKey` and the network name as shown in the following:

```
import { DeployUtil, CLPublicKey } from 'casper-js-sdk';

const deployParams = new DeployUtil.DeployParams(
  CLPublicKey.fromHex(publicKeyHex),
  network_name // 'testnet' | 'mainnet'
);
```

For creating a `session` object, which is `DeployUtil.ExecutableDeployItem`, we need

- The **delegator** and **validator's public keys**
- The **amount** of tokens to delegate/undelegate/redelgate
- The **auction manager contract's hash**

> - The **entry point**.

Reviewer: ```suggestion
- The **entry point**
```

First, create a variable `RuntimeArgs` from the public keys and the amount. We will need to use it below in `session`:

```
import { RuntimeArgs, CLValueBuilder, CLPublicKey } from 'casper-js-sdk';

const args = RuntimeArgs.fromMap({
  delegator: CLPublicKey.fromHex(delegatorPublicKeyHex),
  validator: CLPublicKey.fromHex(validatorPublicKeyHex),
  amount: CLValueBuilder.u512(amountMotes) // in motes
});
```

Second, create a `session` parameter. It is a `Uint8Array` consisting of the `auction manager contract's hash`, the `entry points` and `runtime arguments`, which we previously created.

> The `auction manager contract's hash` will depend on the network you are using. For Casper's mainnet and testnet, the hashes are as follows:

Reviewer: ```suggestion
The `auction manager contract's hash` will depend on the network you are using. For Casper's Mainnet and Testnet, the hashes are as follows:
```

- Mainnet

    `ccb576d6ce6dec84a551e48f0d0b7af89ddba44c7390b690036257a04a3ae9ea`

- Testnet

    `93d923e336b20a4c4ca14d592b60e5bd3fe330775618290104f9beb326db7ae2`

Your `entry point` will depend on which action you are performing, with the following three available:

> - `delegate` - Initial delegation to a validator.

Reviewer: ```suggestion
- `delegate` - Initial delegation to a validator
```

> - `undelegate` - Undelegating tokens from a validator back to the delegator.

Reviewer: ```suggestion
- `undelegate` - Undelegating tokens from a validator back to the delegator
```

> - `redelegate` - Redelegating tokens to a new validator.

Reviewer: ```suggestion
- `redelegate` - Redelegating tokens to a new validator
```

```
import { decodeBase16, DeployUtil } from 'casper-js-sdk';

const session = DeployUtil.ExecutableDeployItem.newStoredContractByHash(
  decodeBase16(auction_manager_contract_hash), // auction manager contract hash
  contractEntryPoint, // auction manager entry point
  args
);
```

> To create the `payment` parameter for the deploy, we need a deploy cost. The actual costs can be pulled from the network [`chainspec`](https://github.com/casper-network/casper-node/blob/release-1.4.8/resources/production/chainspec.toml).

Reviewer: ```suggestion
To create the `payment` parameter for the deploy, we need a deploy cost. The actual costs can be pulled from the network chainspec. Here is the [`chainspec for version 1.4.8`](https://github.com/casper-network/casper-node/blob/release-1.4.8/resources/production/chainspec.toml). You will need the chainspec for the network version you are using.
```

const args = RuntimeArgs.fromMap({
    delegator: CLPublicKey.fromHex(delegatorPublicKeyHex),
    validator: CLPublicKey.fromHex(validatorPublicKeyHex),
    new_validator: CLPublicKey.fromHex(redelegateValidatorPublicKeyHex),
    amount: CLValueBuilder.u512(amountMotes)
})
```
  
#### 2a. Sign the deploy (Casper Signer)

To get the signature, you will need to use `Signer.sign` from the [JavaScript SDK](https://github.com/casper-ecosystem/casper-js-sdk/). It will return `Promise<{ deploy }>`, which is the signed object.

Use `DeployUtil.deployFromJson` to convert the result and sent it to network with:

```
import { Signer, CasperServiceByJsonRPC, DeployUtil } from 'casper-js-sdk';

const casperService = new CasperServiceByJsonRPC(GRPC_URL);
const deployJson = DeployUtil.deployToJson(deploy);
Signer.sign(
    deployJson,
    accountPublicKey,
    recipientPublicKey
).then((signedDeployJson) => {
    const signedDeploy = DeployUtil.deployFromJson(signedDeployJson);
    if (signedDeploy.ok) {
      casperService.deploy(signedDeploy.val! as DeployUtil.Deploy); // sent deploy
    }
}
```

#### 2b. Sign the deploy (Ledger)

> You will need to connect with your`Ledger` first to get the signature. 

Reviewer: ```suggestion
You will need to connect with your `Ledger` first to get the signature. 
```

```
import TransportWebUSB from '@ledgerhq/hw-transport-webusb';
import LedgerApp, { ResponseBase } from '@zondax/ledger-casper';
import { DeployUtil } from 'casper-js-sdk';

const getBipPath = (index: number) => {
  const idx = index.toString();
  return `m/44'/506'/0'/0/${idx}`;
};

const deployBytes = DeployUtil.deployToBytes(deploy) as Buffer;
const transport = await TransportWebUSB.create();
const ledgerApp = new LedgerApp(transport);
const res = await ledgerApp.sign(
    getBipPath(selectedAccountIndex),
    deployBytes
);
```

The Signature will be in a property called `res.signatureRS`.

After that, we can create a signed deploy, 

```
import { DeployUtil, CLPublicKey } from 'casper-js-sdk';

const signedDeploy = DeployUtil.setSignature(
  deploy,
  signatureRS,
  CLPublicKey.fromHex(accountPublicKey)
);
```

We can then send it to a network.

```
casperService.deploy(signedDeploy)
```

### Costs and Minimums

> The following are current costs and minimum amounts, but up-to-date values can be pulled from the network `chainspec`.

Reviewer: ```suggestion
The following are costs and minimum amounts for version 1.4.8, but up-to-date values should be pulled from the network `chainspec`.
```

Transfer Cost: 100,000,000 motes or 0.1 **CSPR**

Delegation Cost: 2,500,000,000 motes or 2.5 **CSPR**

Minimum transfer amount: 2,500,000,000 motes, or 2.5 **CSPR**

Minimum amount required for delegation: 500,000,000,000 motes, or 500 **CSPR**.

### The Delegation Cap

> Casper includes a delegator limit rule, which sets a cap on the number of delegators that a single validator may have at `953`. This is a temporary solution to prevent complications with Casper’s fast sync mechanism - in which high bond counts could break fast sync.

Reviewer: ```suggestion
Casper includes a delegator limit rule, which limits the number of delegators that a single validator may have at `953`. This is a temporary solution to prevent complications with Casper’s fast sync mechanism - in which high bond counts could break fast sync.
```

> Validators with a delegator count at or above `953` at the time of the **1.4.5** upgrade will be grandfathered in, however new delegators will not be able to delegate to any validator until the delegator count for that validator falls below `953`. 

Reviewer: ```suggestion
Validators with a delegator count at or above `953` at the time of the **1.4.5** upgrade were grandfathered in; however, new delegators will not be able to delegate to any validator until the delegator count for that validator falls below `953`. 
```
